services/github/src/handlers/index.js

```diff
@@ -0,0 +1,376 @@
+const stringify = require('fast-json-stable-stringify');
+const crypto = require('crypto');
+const taskcluster = require('taskcluster-client');
+const yaml = require('js-yaml');
+const assert = require('assert');
+const { consume } = require('taskcluster-lib-pulse');
+
+const { deprecatedStatusHandler } = require('./deprecatedStatus');
+const { taskGroupCreationHandler } = require('./taskGroupCreation');
+const { statusHandler } = require('./status');
+const { taskDefinedHandler } = require('./taskDefined');
+const { jobHandler } = require('./job');
+const { rerunHandler } = require('./rerun');
+const { POLICIES } = require('./policies');
+
+/**
+ * Create handlers
+ */
+class Handlers {
+  constructor(options) {
+    const {
+      rootUrl,
+      credentials,
+      monitor,
+      reference,
+      jobQueueName,
+      deprecatedResultStatusQueueName,
+      deprecatedInitialStatusQueueName,
+      resultStatusQueueName,
+      initialStatusQueueName,
+      rerunQueueName,
+      intree,
+      context,
+      pulseClient,
+    } = options;
+
+    assert(monitor, 'monitor is required for statistics');
+    assert(reference, 'reference must be provided');
+    assert(rootUrl, 'rootUrl must be provided');
+    assert(intree, 'intree configuration builder must be provided');
+    this.rootUrl = rootUrl;
+    this.credentials = credentials;
+    this.monitor = monitor;
+    this.reference = reference;
+    this.intree = intree;
+    this.connection = null;
+    this.deprecatedResultStatusQueueName = deprecatedResultStatusQueueName;
+    this.resultStatusQueueName = resultStatusQueueName;
+    this.jobQueueName = jobQueueName;
+    this.rerunQueueName = rerunQueueName;
+    this.deprecatedInitialStatusQueueName = deprecatedInitialStatusQueueName;
+    this.initialStatusQueueName = initialStatusQueueName;
+    this.context = context;
+    this.pulseClient = pulseClient;
+
+    this.handlerComplete = null;
+    this.handlerRejected = null;
+
+    this.commentHashCache = [];
+
+    this.jobPq = null;
+    this.resultStatusPq = null;
+    this.deprecatedResultStatusPq = null;
+    this.initialTaskStatusPq = null;
+    this.deprecatedInitialStatusPq = null;
+    this.rerunPq = null;
+
+    this.queueClient = null;
+  }
+
+  /**
+   * Set up the handlers.
+   */
+  async setup(options = {}) {
+    assert(!this.jobPq, 'Cannot setup twice!');
+    assert(!this.resultStatusPq, 'Cannot setup twice!');
+    assert(!this.initialTaskStatusPq, 'Cannot setup twice!');
+    assert(!this.deprecatedResultStatusPq, 'Cannot setup twice!');
+    assert(!this.deprecatedInitialStatusPq, 'Cannot setup twice!');
+    assert(!this.rerunPq, 'Cannot setup twice!');
+
+    // This is a powerful Queue client without scopes to use throughout the handlers for things
+    // where taskcluster-github is acting of its own accord
+    // Where it is acting on behalf of a task, use this.queueClient.use({authorizedScopes: scopes}).blahblah
+    // (see this.createTasks for example)
```

Comment:
I think this is an important point -- `this.queueClient.rerunTask()` will use tc-github's scopes, which means it can rerun just about any task.  Other handlers have things like
https://github.com/taskcluster/taskcluster/blob/3eddaee81983d2f631952904298bbbf637afc272/services/github/src/handlers.js#L460-L462
in this case, `scopes` would probably be defined as something like `["assume:repo:github.com/<org>/<repo>:rerun"]`.

Reply:
Thanks! Will apply this scope